library/Translate/src/Juvix/FrontendContextualise/Contextify/Sexp.hs

```diff
@@ -0,0 +1,223 @@
+  | Just Sexp.N {atomNum} <- Sexp.atomFromT i,
+    Just atomName <- eleToSymbol n =
+    let prec =
+          Context.Pred
+            if  | Sexp.isAtomNamed inf "infix" ->
+                  Context.NonAssoc
+                | Sexp.isAtomNamed inf "infixl" ->
+                  Context.Left
+                | Sexp.isAtomNamed inf "infixr" ->
+                  Context.Right
+                | otherwise -> error "malformed declaration"
+            (fromIntegral atomNum)
+     in pure $ case Context.extractValue <$> Context.lookup (pure atomName) ctx of
+          Just (Context.Def d) ->
+            Type.PS
+              { ctxS =
+                  ctx
+                    |> Context.add
+                      (NameSpace.Pub atomName)
+                      (Context.Def (d {Context.defPrecedence = prec})),
+                opensS = [],
+                modsDefinedS = []
+              }
+          _ ->
+            Type.PS
+              { ctxS =
+                  Context.add
+                    (NameSpace.Pub atomName)
+                    (Context.Information [Context.Prec prec])
+                    ctx,
+                opensS = [],
+                modsDefinedS = []
+              }
+declare _ _ = error "malformed declare"
+
+-- TODO ∷ don't ignore the assoc information given with type'
+type' :: Sexp.T -> Context.T Sexp.T Sexp.T Sexp.T -> IO Type.PassSexp
+type' t@(assocName Sexp.:> _ Sexp.:> dat) ctx
+  | Just name <- eleToSymbol (Sexp.car assocName) =
+    let constructors = collectConstructors dat
+        addSum con =
+          Context.Sum Nothing name
+            |> Context.SumCon
+            |> Context.add (NameSpace.Pub con)
+        newCtx = foldr addSum ctx constructors
+     in pure $
+          Type.PS
+            { ctxS = Context.add (NameSpace.Pub name) (Context.TypeDeclar t) newCtx,
+              opensS = [],
+              modsDefinedS = []
+            }
+type' _ _ = error "malformed type"
+
+open :: Sexp.T -> Context.T Sexp.T Sexp.T Sexp.T -> IO Type.PassSexp
+open (Sexp.List [mod]) ctx
+  | Just Sexp.A {atomName} <- Sexp.atomFromT mod =
+    pure $
+      Type.PS
+        { ctxS = ctx,
+          opensS = [atomName],
+          modsDefinedS = []
+        }
+open _ _ = error "malformed open"
+
+-- TODO ∷ why is the context empty?
+-- we should somehow note what lists are in scope
+
+-- TODO ∷
+-- - once we have type checking, rely on that
+-- - for dep types where inference is undecidable, force signature
+-- - for functions like (f x) where that evals to a module, where it
+--   is dependent but decidable, force signature?
+
+-- The NameSymbol.T return returns back all record names that are found
+-- we don't return a map, as we can't do opens in a record
+
+-- | decideRecordOrDef tries to figure out
+-- if a given defintiion is a record or a definition
+decideRecordOrDef ::
+  Sexp.T ->
+  Symbol ->
+  NameSymbol.T ->
+  Context.Precedence ->
+  Maybe Sexp.T ->
+  IO (Type.DefinitionSexp, [NameSymbol.T])
```

Comment:
is there a reason this is in `IO` instead of `MonadIO` ?

Reply:
Probably so I dont' have to lift in emptyRecord and other such functions.

Reply:
hmm let's make sure this is standardised?

Reply:
I moved it to MonadIO... all the functions inside are stil IO except for that one, but since we really only want to export a few, it would make sense to have exported things be in MonadIO?